### DISCUSSION TITLE
XBRL and Financials rewrite

### DISCUSSION THREAD

**Discussion by dgunning** [dgunning/edgartools#82] [General]
I pushed a rewrite of the XBRL Parser and Financials. The main features are

- Generate financials statements exactly as published by companies
- Extract all statements not just financials e.g. APPL 10-K has 78 statements
- Easier querying of XBRL data

Here is a link to a notebook with some of the features
[Reading Data from XBRL](https://github.com/dgunning/edgartools/blob/main/notebooks/Reading-Data-From-XBRL.ipynb)

Feedback requested

**Comment by dgunning** [comment]
The new XBRL API allows you to get a statement by name or index
<img width="847" alt="consolidated-statement" src="https://github.com/user-attachments/assets/4a8c3413-5e29-42f3-acf8-284fc98b001c">

**Comment by amitgandhinz** [comment]
income_statement.get_concept("us-gaap:RevenueFromContractWithCustomerExcludingAssessedTax")

When I try this on TSLA's latest 10-Q (Q2) I get an error 

```
/edgar/xbrl/xbrldata.py", line 535, in get_concept
    name=results.concept.iloc[0],
AttributeError: 'str' object has no attribute 'iloc'
```

**Reply by amitgandhinz** [reply]
created an issue for this here: https://github.com/dgunning/edgartools/issues/88

**Comment by tjf2007** [comment]
This is really excellent, Dwight, thank you. 

**Three questions for you:**
1. Is there an easy way to get the full history of financial statements and for quarterly intervals?
2. How do you think about dealing with the **re-statements** point-in-time?
3. For Quarterly financial statement data with companies that publish Q1, Q2, Q3, and FY values, do you derive what the Q4 value is (e.g., `Q4 = FY - [Q1 + Q2 + Q3]`)? An example is IBM

**Comment by tjf2007** [comment]
What do you think is the best way to get the explicit reference period for the Financials information?


For example:
Let's say I want to see the point-in-time snapshot of IBM's balance sheet, as of `2010-02-23` (the acceptance date for their `10-K` submitted with ain: `'0001047469-10-001151'`):


```
ain = '0001047469-10-001151' # IBM 10-K, accepted on 2011-02-22T13:08:33.000Z
filing = get_by_accession_number(ain)
xbrl_data: XBRLData = XBRLData.extract(filing)
financials = Financials(xbrl_data)
bs = financials.get_balance_sheet()
print(bs.periods). # ['2009', '2008']```

What I'd really like to see is the "as of" reference period, so instead of ['2009', '2008'], it would be the actual fiscal period as of dates, 12/31/2009 and 12/31/2008. That will be more important when stitching together quarterly data, too.

**Reply by dgunning** [reply]
Gotcha. You would prefer to see the dates rather that 2023, 2024 as the column headers?
I could change it to that.
